app/models/application_form.rb

```diff
@@ -252,6 +252,14 @@ def can_add_more_choices?
     choices_left_to_make.positive?
   end
 
+  def courses_not_yet_open
+    @courses_not_yet_open ||= application_choices.each.with_object([]) { |choice, array| array << choice.course if choice.course.applications_open_from > Time.zone.today }
+  end
+
+  def full_courses
+    @full_courses ||= application_choices.each.with_object([]) { |choice, array| array << choice.course if choice.course_option.no_vacancies? }
+  end
```

Comment:
imho these methods dilute the model a bit. could each of them go into a separate query class?